src/flat.rs

```diff
@@ -0,0 +1,1139 @@
+        self.format.shrink_to(width, height, channels)
+    }
+}
+
+/// A flat buffer that can be used as an image view.
+///
+/// This is a nearly trivial wrapper around a buffer but at least sanitizes by checking the buffer
+/// length first and constraining the pixel type.
+///
+/// Note that this does not eliminate panics as the `AsRef<[T]` implementation of `Buffer` may be
+/// unreliable, i.e. return different buffers at different times. This of course is a non-issue for
+/// all common collections where the bounds check once must be enough.
+#[derive(Clone, Debug)]
+pub struct View<Buffer, P: Pixel> 
+where 
+    Buffer: AsRef<[P::Subpixel]> 
+{
+    inner: FlatSamples<Buffer>,
+    phantom: PhantomData<P>,
+}
+
+/// A mutable owning version of a flat buffer.
+///
+/// While this wraps a buffer similar to `ImageBuffer`, this is mostly intended as a utility. The
+/// library endorsed normalized representation is still `ImageBuffer`. Also, the implementation of
+/// `AsMut<[P::Subpixel]>` must always yield the same buffer. Therefore there is no public way to
+/// construct this with an owning buffer.
+#[derive(Clone, Debug)]
+pub struct ViewMut<Buffer, P: Pixel> 
+where 
+    Buffer: AsMut<[P::Subpixel]> 
+{
+    inner: FlatSamples<Buffer>,
+    phantom: PhantomData<P>,
+}
+
+/// Denotes invalid flat sample buffers when trying to convert to stricter types.
+///
+/// The biggest use case being `ImageBuffer` which expects closely packed
+/// samples in a row major matrix representation. But this error type may be
+/// resused for other import functions. A more versatile user may also try to
+/// correct the underlying representation depending on the error variant.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum Error {
+    /// The represented image was too large.
+    ///
+    /// The optional value denotes a possibly accepted maximal bound.
+    TooLarge,
+
+    /// The represented image can not use this representation.
+    ///
+    /// The normalized form that would be accepted.
+    NormalFormRequired(NormalForm),
+
+    /// The color format did not match the channel count.
+    ///
+    /// In some cases you might be able to fix this by lowering the reported pixel count of the
+    /// buffer without touching the strides.
+    ///
+    /// In very special circumstances you *may* do the opposite. This is **VERY** dangerous but not
+    /// directly memory unsafe although that will likely alias pixels. One scenario is when you
+    /// want to construct an `Rgba` image but have only 3 bytes per pixel and for some reason don't
+    /// care about the value of the alpha channel even though you need `Rgba`.
+    WrongColor(ColorType),
+}
+
+/// Different normal forms of buffers.
+///
+/// A normal form is an unaliased buffer with some additional constraints.  The `ÌmageBuffer` uses
+/// row major form with packed samples. 
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum NormalForm {
```

Comment:
I'd rather have offered this publically with and additional `normalize` function in the spirit of `slice::sort` or `slice::reverse`. That way, one could import a general buffer and normalize it at a convenient time before calling a never-failing `into_buffer`, instead of having to copy a `View` into a newly allocated `ImageBuffer` or having to handroll this.